### PR TITLE
Stcor 276

### PR DESCRIPTION
**Pull request by anton-plotnikov**
Forgot Username page component implementation

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -113,6 +113,7 @@
     "react-apollo": "^2.1.3",
     "react-cookie": "^2.1.4",
     "react-dom": "^16.3.0",
+    "react-flexbox-grid": "^2.1.2",
```
Comment (maximdidenkoepam) [inline package.json]: Why do we need this dependency?

File: src/components/ForgotUserName/ForgotUserNameForm.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  Headline,
+  Row,
+  Col,
+} from '@folio/stripes-components';
+
+import OrganizationLogo from '../OrganizationLogo/OrganizationLogo';
+import AuthErrorsContainer from '../AuthErrorsContainer/index';
+
+import formStyles from '../Login/AuthForm.css';
+import errorContainerStyles
```
Comment (maximdidenkoepam) [inline src/components/ForgotUserName/ForgotUserNameForm.js]: move from to the same line as import

File: src/components/ForgotUserName/ForgotUserNameCtrl.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    if (this.state.userDoesNotExist === true) {
+      return (
+        <ForgotUserNameForm
+          onSubmit={this.handleSubmit}
+          isNotValid={this.state.isNotValidEmail}
+          hasErrorsContainer={this.state.hasErrorsContainer}
+        />
+      );
+    }
+    return <Redirect to={{
```
Comment (maximdidenkoepam) [inline src/components/ForgotUserName/ForgotUserNameCtrl.js]: The component is called` ForgotUserNameCtrl.` I think that in this case by default `ForgotUserNameForm` should be returned. I suggest to do invert if statement and inside of it you should do redirect. Also we may need to rename `userDoesNotExist`. And also I would use state destructuring on the top of the method.

File: src/helpers/validateEmail/validateEmail.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import normalizeInput from '../normalizeInput';
+
+export default email => {
+// eslint-disable-next-line no-useless-escape
```
Comment (maximdidenkoepam) [inline src/helpers/validateEmail/validateEmail.js]: I assume that `validateEmail` should validate email, not the `phoneNumber`? I suggest to split it in to two separate validators and use Facade to combine it

File: src/components/ForgotUserName/ForgotUserNameCtrl.js (hunk 3 above)
Comment (anton-plotnikov) [reply src/components/ForgotUserName/ForgotUserNameCtrl.js]: Suggest that userDoesNotExist is a logical initial state of the page: we do not know it beforehand

File: src/components/ForgotUserName/ForgotUserNameCtrl.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      {
+        isNotValidEmail: false,
+        userDoesNotExist: true,
+        hasErrorsContainer: false,
+      }
+    );
+  };
+
+  render() {
+    if (this.state.userDoesNotExist === true) {
```
[the remark this answers is not in the document]
Comment (anton-plotnikov) [reply src/components/ForgotUserName/ForgotUserNameCtrl.js]: Maybe we can leave the comparison to be more explicit and readable?

File: src/RootWithIntl.js [hunk 6]
```diff
@@ -30,6 +30,8 @@ import {
   Login,
   OverlayContainer,
   CreateResetPassword,
+  ForgotUserNameCtrl,
+  CheckEmailStatusPage
```
Comment (VictorSoroka1) [inline src/RootWithIntl.js]: Add trailing comma.

File: src/components/CheckEmailStatusPage/CheckEmailStatusPage.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
+import React, { Fragment } from 'react';
+import { withRouter } from 'react-router-dom';
+import { FormattedMessage } from 'react-intl';
+import PropTypes from 'prop-types';
+
+import Headline from '@folio/stripes-components/lib/Headline';
+import OrganizationLogo from '../OrganizationLogo/OrganizationLogo';
+
+import styles from './CheckEmailStatusPage.css';
+import validateEmail from '../../validators/validateEmail/validateEmail';
```
Comment (VictorSoroka1) [inline src/components/CheckEmailStatusPage/CheckEmailStatusPage.js]: Move it above the styles to the lines with js files.

File: src/components/CheckEmailStatusPage/CheckEmailStatusPage.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            <FormattedMessage id={`${labelNamespace}.caution.email`} />
+          </Headline>
+        </div>
+      </div>
+    </Fragment>
+  );
+};
+
+CheckEmailStatusPage.propTypes = {
+  location:PropTypes.object.isRequired,
```
Comment (VictorSoroka1) [inline src/components/CheckEmailStatusPage/CheckEmailStatusPage.js]: As you are using 

```
    location: {
      state: {
        userEmail,
      },
    },
```
add PropTypes for these as well.

File: src/components/ForgotUserName/ForgotUserNameCtrl.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+      return <Redirect to={{
+        pathname: '/check-email',
+        state: { userEmail },
+      }}
+      />;
+    }
+
+    return (
+      <ForgotUserNameForm
+        onSubmit={this.handleSubmit}
```
Comment (VictorSoroka1) [inline src/components/ForgotUserName/ForgotUserNameCtrl.js]: Move handlers to the bottom.

File: src/components/ForgotUserName/ForgotUserNameForm.js [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+import {
+  TextField,
+  Button,
+  Headline,
+  Row,
+  Col,
+} from '@folio/stripes-components';
+
+import OrganizationLogo from '../OrganizationLogo/OrganizationLogo';
+import AuthErrorsContainer from '../AuthErrorsContainer/index';
```
Comment (VictorSoroka1) [inline src/components/ForgotUserName/ForgotUserNameForm.js]: I suppose `index` is excess at the end.

File: src/components/ForgotUserName/ForgotUserNameForm.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    } = this.props;
+
+    return (
+      <Fragment>
+        <div className={formStyles.wrap}>
+          <div className={formStyles.centered}>
+            <OrganizationLogo />
+            <Form
+              className={formStyles.form}
+              onSubmit={handleSubmit(onSubmit)}
```
Comment (VictorSoroka1) [inline src/components/ForgotUserName/ForgotUserNameForm.js]: Move handlers to the bottom.

File: src/components/ForgotUserName/ForgotUserNameForm.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+                marginBottom0
+                data-test-submit
+              >
+                <FormattedMessage id="stripes-core.button.continue" />
+              </Button>
+              <div
+                className={errorContainerStyles.authErrorsWrapper}
+                data-test-errors
+              >
+                { hasErrorsContainer &&
```
Comment (VictorSoroka1) [inline src/components/ForgotUserName/ForgotUserNameForm.js]: Please remove space before `hasErrorsContainer `.

File: src/components/ForgotUserName/ForgotUserNameForm.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                <Row center="xs">
```
Comment (VictorSoroka1) [inline src/components/ForgotUserName/ForgotUserNameForm.js]: Add indentation.

File: src/validators/validateEmail/validateEmail.js [hunk 14]
```diff
@@ -0,0 +1,5 @@
+export default email => {
+// eslint-disable-next-line no-useless-escape
+  const emailRegExp = /^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i;
+  return emailRegExp.test(email);
```
Comment (VictorSoroka1) [inline src/validators/validateEmail/validateEmail.js]: Add an empty line before `return` statement.

File: src/validators/validateEmail/validateEmail.js [hunk 15]
```diff
@@ -0,0 +1,5 @@
+export default email => {
+// eslint-disable-next-line no-useless-escape
```
Comment (VictorSoroka1) [inline src/validators/validateEmail/validateEmail.js]: Align indentation with other code.

File: src/validators/validatePhoneNumber/validatePhoneNumber.js [hunk 16]
```diff
@@ -0,0 +1,4 @@
+export default phoneNumber => {
+  const phoneRegExp = /^(\d+[.-]*)+\d+$/;
+  return phoneRegExp.test(phoneNumber);
```
Comment (VictorSoroka1) [inline src/validators/validatePhoneNumber/validatePhoneNumber.js]: Add an empty line before `return` statement.

File: test/bigtest/interactors/ErrorsContainer.js [hunk 17]
```diff
@@ -0,0 +1,12 @@
+import {
+  interactor,
+  scoped,
+} from '@bigtest/interactor';
+
+@interactor
+class ErrorsContainerInteractor {
+  static defaultScope = ('[data-test-errors]');
+  errorsContainer = scoped('[class^="AuthErrorsContainer--"]');
```
Comment (VictorSoroka1) [inline test/bigtest/interactors/ErrorsContainer.js]: Add an empty line after `static` statement.

File: test/bigtest/interactors/ForgotUsername.js [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+
+import TextFieldInteractor
+  from '@folio/stripes-components/lib/TextField/tests/interactor';
+import SubmitButtonInteractor from './SubmitButton';
+import ErrorsContainerInteractor from './ErrorsContainer';
+import StatusPageInteractor from './StatusPage';
+
+@interactor
+class ForgotUsernameInteractor {
+  static defaultScope = ('form[class^="form--"]');
```
Comment (VictorSoroka1) [inline test/bigtest/interactors/ForgotUsername.js]: Add an empty line after static statement.

File: test/bigtest/interactors/StatusPage.js [hunk 19]
```diff
@@ -0,0 +1,14 @@
+import {
+  interactor,
+  scoped,
+} from '@bigtest/interactor';
+
+@interactor
+class StatusPageInteractor {
+  static defaultScope = ('[data-test-status-page]');
```
Comment (VictorSoroka1) [inline test/bigtest/interactors/StatusPage.js]: Add an empty line after static statement.

File: test/bigtest/interactors/SubmitButton.js [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+{
+  interactor,
+  is,
+} from '@bigtest/interactor';
+import ButtonInteractor
+  from '@folio/stripes-components/lib/Button/tests/interactor';
+
+export default @interactor
+class SubmitButtonInteractor {
+  static defaultScope = ('button[data-test-submit]');
```
Comment (VictorSoroka1) [inline test/bigtest/interactors/SubmitButton.js]: Add an empty line after static statement.

File: test/bigtest/tests/forgotUsername-test.js [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+  const {
+    inputField,
+    submitButton,
+    submitButton: { button },
+    mainHeading,
+    callToActionParagraph,
+    errorsWrapper,
+    errorsWrapper: { errorsContainer },
+    statusPage,
+    statusPage:
```
Comment (VictorSoroka1) [inline test/bigtest/tests/forgotUsername-test.js]: Why not write it as

```suggestion
    statusPage: {
      heading: statusPageHeading,
      notificationParagraph,
      cautionParagraph,
    },
```

File: test/bigtest/tests/forgotUsername-test.js [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+          heading: statusPageHeading,
+          notificationParagraph,
+          cautionParagraph,
+        },
+  } = forgotUsernamePage;
+  const invalidInput = 'asdfgh12345';
+  const nonExistingRecord = '12345';
+  const existingRecord = '127-699-8925';
+
+  describe('Forgot form text input field tests', () => {
```
Comment (VictorSoroka1) [inline test/bigtest/tests/forgotUsername-test.js]: Start message for `describe` and `it` with lowercase.

File: test/bigtest/tests/forgotUsername-test.js [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+    it('Should have a disabled "Continue" button by default', () => {
+      expect(submitButton.isDisabled).to.be.true;
+    });
+  });
+
+  describe('Forgot form submit button test after filling the input', () => {
+    beforeEach(async () => {
+      await inputField.fillInput(invalidInput);
+    });
+    it('Should have an enabled submit button', () => {
```
Comment (VictorSoroka1) [inline test/bigtest/tests/forgotUsername-test.js]: Add an empty line between `beforeEach` and `it` and between individual `it`. Be consistent with that.

File: test/bigtest/tests/forgotUsername-test.js [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+    in english translation`, () => {
+      expect(mainHeading.text).to.equal(translations['label.forgotUsername']);
+    });
+
+    it('Should display the paragraph', () => {
+      expect(callToActionParagraph.isPresent).to.be.true;
+    });
+
+    it(`Should have the paragraph content equal to forgot username or password
+    call to action label in english translation`, () => {
```
Comment (VictorSoroka1) [inline test/bigtest/tests/forgotUsername-test.js]: Add indentation. Handle other similar cases.

File: test/bigtest/tests/forgotUsername-test.js [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+    });
+  });
+
+  describe('Error container initial behaviour', () => {
+    describe('Forgot form error container integration test', () => {
+      it('Should display the error container wrapper', () => {
+        expect(errorsWrapper.isPresent).to.be.true;
+      });
+
+      it('Should not display the error container', () => {
```
Comment (VictorSoroka1) [inline test/bigtest/tests/forgotUsername-test.js]: I think it should be `it.always`. Read about it here [here](https://github.com/bigtestjs/convergence). Check other similar cases.

File: test/bigtest/tests/statusPage-test.js [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+
+import translations from '../../../translations/stripes-core/en';
+import setupApplication from '../helpers/setup-application';
+import StatusPageInteractor from '../interactors/StatusPage';
+
+describe('Forgot username form test', () => {
+  setupApplication({ disableAuth: false });
+
+  beforeEach(function () {
+    this.visit(
```
Comment (VictorSoroka1) [inline test/bigtest/tests/statusPage-test.js]:
```suggestion
    this.visit({
      pathname: '/check-email',
      state: { userEmail: '127-699-8925' },
    });
```

File: src/components/CheckEmailStatusPage/CheckEmailStatusPage.js [hunk 27]
```diff
@@ -0,0 +1,72 @@
+import React, { Fragment } from 'react';
+import { withRouter } from 'react-router-dom';
+import { FormattedMessage } from 'react-intl';
+import PropTypes from 'prop-types';
+
+import Headline from '@folio/stripes-components/lib/Headline';
```
Comment (VictorSoroka1) [inline src/components/CheckEmailStatusPage/CheckEmailStatusPage.js]: Let's add an additional empty line after imports from `@folio`.

File: src/components/ForgotUserName/ForgotUserNameForm.js [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+                  <Col xs={12}>
+                    <AuthErrorsContainer
+                      errors={
+                        isValid
```
Comment (VictorSoroka1) [inline src/components/ForgotUserName/ForgotUserNameForm.js]:
```suggestion
                        [{ code: isValid ? 'unable.locate.account' : 'email.invalid' }]
```

File: test/bigtest/tests/forgotUsername-test.js [hunk 29]
```diff
@@ -0,0 +1,190 @@
+import { describe, it, beforeEach } from '@bigtest/mocha';
+import { expect } from 'chai';
+import translations from '../../../translations/stripes-core/en';
```
Comment (VictorSoroka1) [inline test/bigtest/tests/forgotUsername-test.js]: Let's add additional space after third-party imports.

File: src/components/ForgotUserName/ForgotUserNameCtrl.js (hunk 9 above)
Comment (anton-plotnikov) [reply src/components/ForgotUserName/ForgotUserNameCtrl.js]: Existing ESLint config requires render method to be at the bottom

File: src/RootWithIntl.js [hunk 30]
```diff
@@ -128,6 +130,16 @@ class RootWithIntl extends React.Component {
                         component={<CookiesProvider><SSOLanding stripes={stripes} /></CookiesProvider>}
                         key="sso-landing"
                       />
+                      <TitledRoute
+                        name="Forgot username"
+                        path="/forgot-username"
+                        component={<ForgotUserNameCtrl stripes={stripes} />}
+                      />
+                      <TitledRoute
+                        name="Check email"
```
[the remark this answers is not in the document]
Comment (anton-plotnikov) [reply src/RootWithIntl.js]: Yes, sure

File: src/RootWithIntl.js [hunk 31]
```diff
@@ -128,6 +130,16 @@ class RootWithIntl extends React.Component {
                         component={<CookiesProvider><SSOLanding stripes={stripes} /></CookiesProvider>}
                         key="sso-landing"
                       />
+                      <TitledRoute
+                        name="Forgot username"
```
[the remark this answers is not in the document]
Comment (anton-plotnikov) [reply src/RootWithIntl.js]: Yes, sure

**Pull request merged**